src/main/scala/is/hail/expr/ir/MatrixValue.scala

```diff
@@ -188,68 +203,75 @@ case class MatrixValue(
-
-        val colRegion = ctx.freshRegion
-        val colRVB = new RegionValueBuilder(colRegion)
-        val colsType = TArray(localColType, required = true)
-        colRVB.start(colsType)
-        colRVB.addAnnotation(colsType, localColValues.value)
-        val colsOffset = colRVB.end()
-
-        it.flatMap { rv =>
-          val gsOffset = fullRowType.loadField(rv, localEntriesIndex)
-          (0 until localNCols).iterator
-            .filter { i =>
-              localEntriesType.isElementDefined(rv.region, gsOffset, i)
-            }
-            .map { i =>
-              rv2b.clear()
-              rv2b.start(resultStruct)
-              rv2b.startStruct()
-
-              var j = 0
-              while (j < fullRowType.size) {
-                if (j != localEntriesIndex)
-                  rv2b.addField(fullRowType, rv, j)
-                j += 1
+      val localSortedColValues = sortedColValues.broadcast
+      val localSortedColsToOldIdx = sortedColsToOldIdx.broadcast
+
+      rvd.constrainToOrderedPartitioner(rvd.partitioner.strictify).boundary
+        .mapPartitionsPreservesPartitioning(typ.entriesTableType.rvdType.copy(key = typ.rowKey),
+          { (ctx, it) =>
+            val rv2b = ctx.rvb
+            val rv2 = RegionValue(ctx.region)
+
+            val colRegion = ctx.freshRegion
+            val colRVB = new RegionValueBuilder(colRegion)
+            val colsType = TArray(localColType, required = true)
+            colRVB.start(colsType)
+            colRVB.addAnnotation(colsType, localSortedColValues.value)
+            val colsOffset = colRVB.end()
+
+            val rowBuffer = new RegionValueArrayBuffer(fullRowType, ctx.freshRegion)
+
+            val colsNewToOldIdx = localSortedColsToOldIdx.value.asInstanceOf[IndexedSeq[Int]]
+
+            OrderedRVIterator(localOrvdType, it, ctx).staircase.flatMap { step =>
+              rowBuffer.clear()
+              rowBuffer ++= step
+              (0 until localNCols).iterator.flatMap { i =>
+                rowBuffer.iterator
+                  .filter { row =>
+                    localEntriesType.isElementDefined(row.region, fullRowType.loadField(row, localEntriesIndex), colsNewToOldIdx(i))
+                  }.map { row =>
+                    rv2b.clear()
+                    rv2b.start(resultStruct)
+                    rv2b.startStruct()
+
+                    var j = 0
+                    while (j < fullRowType.size) {
+                      if (j != localEntriesIndex)
+                        rv2b.addField(fullRowType, row, j)
+                      j += 1
+                    }
+
+                    rv2b.addAllFields(localColType, colRegion, colsType.loadElement(colRegion, colsOffset, i))
+                    rv2b.addAllFields(
+                      localEntryType,
+                      row.region,
+                      localEntriesType.elementOffsetInRegion(row.region, fullRowType.loadField(row, localEntriesIndex), colsNewToOldIdx(i)))
```

Comment:
I feel like we're supposed to use `loadElement` but it doesn't matter since the element type is a struct. @catoverdrive thoughts?

Reply:
I'll approve pending amanda's response.

Reply:
that seems correct (should be `loadElement` but since it's a struct there's no effective difference).

Reply:
let's change to loadElement @patrick-schultz

Reply:
I'm off to soccer, but if you want to get someone else to approve after that change that's ok with me.